PV - Enable logout for users with 2fa disabled when the Force 2fa feature flag is turned on

pr33thi opened the pull request:
Ticket: https://manage.dimagi.com/default.asp?270986

Issue: If this feature flag is enabled for a user who doesn't have 2fa activated, the only pages accessible to them are the account settings pages (where they can enable 2fa). That means that they can't even log out unless they have 2fa enabled.

Fix: Added paths that begin with '/accounts/' to the list of pages that don't give back a 403 error, because the logout page is accounts/logout.

Also updated the tests and got rid of an unmatched closing tag in an unrelated html file

pr33thi committed:
Add 'acounts' url exception to 403 error response

pr33thi committed:
Remove unmatched closing tag from unrelated page

pr33thi committed:
Add unit tests for logout page

pr33thi committed:
Fix indentation

proteusvacuum commented on corehq/apps/users/middleware.py:
```diff
@@ -80,7 +80,8 @@ def process_view(self, request, view_func, view_args, view_kwargs):
                 and not request.user.is_verified()
-                and not request.path.startswith('/account/')):
+                and not request.path.startswith('/account/')
+                and not request.path.startswith('/accounts/')):
```
err... lol. When the `2FA` urls were made they didn't follow the same pattern as the original account ones (which is why this worked in the first place).

I think this would be a good place to use a decorator on the views that shouldn't have superuser 2FA, instead of hard coding the urls here. I think we discussed this earlier too. Something like `@disable_superuser_twofactor`, which would set an attribute on the view function which you could check here instead. 

This change also seems like it is too broad for the feature requested. URLs that start with `accounts` also include things like changing your password, which you shouldn't be allowed to do without enabling 2FA.

pr33thi replied:
Ohh, I thought that decorator logic was part of the PR that was getting QA'd, so I figured this would be a temporary fix until then (which was why I didn't get rid of the request.path.startswith stuff in my first PR of this stuff). Is that not the case? If so, Ill make it for this pr

proteusvacuum replied:
The decorator in that PR does something else (disables 2-FA entirely). This decorator would follow the same pattern, but would only disable the superuser 2FA thing.

pr33thi replied:
But the fix on the PR that preceded this was to allow 2FA to be enforced for all users (not just superusers and staff)

Do you mean there should be a decorator that specifically disables that feature flag on certain views?

proteusvacuum replied:
Ah, I think I see what you mean. I suppose you could use the same decorator and check for it here, is that what you are proposing? i think that makes sense. However, do you know why this check happens in middleware? Or why the other 2FA check doesn't happen in middleware? Maybe there is a way we can consolidate both?

pr33thi replied:
Consolidating sounds reasonable, I'll look into that when I have time / am not on interrupt anymore

pr33thi replied:
Oh yeah, should I wait for your changes to be merged in so I can use that decorator without causing too many conflicts?

pr33thi committed:
Merge branch 'master' into pv/enable-logout-270986

pr33thi committed:
Merge branch 'master' into pv/enable-logout-270986

pr33thi committed:
Remove the accounts url exclusion from this toggle

pr33thi committed:
Consolidate 2FA requirement in with the login_required decorator (but keep it as a feature flag)

pr33thi committed:
Remove unused import

pr33thi committed:
Remove redundant parens

proteusvacuum commented on corehq/apps/domain/decorators.py:
```diff
@@ -314,6 +315,11 @@ def _inner(request, domain, *args, **kwargs):
+    if toggles.TWO_FACTOR_SUPERUSER_ROLLOUT.enabled(couch_user.username):
+        require_2fa = True
+    if require_2fa:
+        return not getattr(view_func, 'two_factor_exempt', False)
```
can you just stick this in the previous `if` statement? I'm not sure what the `require_2fa` variable adds here.

pr33thi replied:
Moved around here:
https://github.com/dimagi/commcare-hq/pull/19570/commits/9ba1f0dba28bb6befaf561594655cadebf644f48

proteusvacuum commented on corehq/apps/settings/views.py:
```diff
@@ -205,9 +215,11 @@ def form_actions(self):
 
     @property
     @memoized
+    @two_factor_exempt
```
You should only need the decorator on the `dispatch` method I think.

pr33thi replied:
Fixed here:
https://github.com/dimagi/commcare-hq/pull/19570/commits/6fa949855daef5a6098e132b6e4d2382d812944a

proteusvacuum reviewed (changes requested):
Nice cleanup!

pr33thi committed:
Add test for feature flag to make two_factor_required always return true

pr33thi committed:
Delete unneeded var and if statement

pr33thi committed:
Remove all @two_factor_exempt decorators not on the dispatch function

pr33thi committed:
Cleanup tests, remove unneeded request from setup

pr33thi committed:
Make create_request a classmethod

pr33thi committed:
Made dummy view func text instead of arbitrary number

pr33thi committed:
Make create_request a classmethod

proteusvacuum commented on corehq/apps/domain/test_decorators.py:
```diff
@@ -0,0 +1,56 @@
+        request = self.request_2fa_disabled
+        self.enable_two_factor_for_user(request)
+        two_factor_required_bool = _two_factor_required(view_func, self.domain, request.couch_user)
+        self.assertEqual(two_factor_required_bool, True)
```
you can use `self.assertTrue` instead

proteusvacuum commented on corehq/apps/domain/test_decorators.py:
```diff
@@ -0,0 +1,56 @@
+
+    @flag_enabled('TWO_FACTOR_SUPERUSER_ROLLOUT')
+    def test_two_factor_required_with_feature_flag(self):
+        view_func=5
```
What is `5`?

proteusvacuum commented on corehq/apps/domain/test_decorators.py:
```diff
@@ -0,0 +1,56 @@
+
+    @classmethod
+    def enable_two_factor_for_user(cls, request):
+        request.user.otp_device = "test_device"
```
how is this enabling 2fa for the user? Not sure I understand what this is doing...

proteusvacuum commented on corehq/apps/domain/test_decorators.py:
```diff
@@ -0,0 +1,56 @@
+        request.couch_user = CouchUser()
+        # Login
+        assert Client().login(username=username, password=password)
+        return request
```
The way you've written your tests, I don't think you need any of this... you can just pass in the `CouchUser` to the function you are testing.

pr33thi replied:
My new tests need request.couch_user, so I changed this to just initialize a request and creat a couch user, unneeded stuff is deleted here:

https://github.com/dimagi/commcare-hq/pull/19570/commits/474935b1ae27265ae264bcc16efb7a2fa2102efd

proteusvacuum commented on corehq/apps/domain/test_decorators.py:
```diff
@@ -0,0 +1,54 @@
+from corehq.apps.domain.decorators import _two_factor_required
+
+
+class TestDecorators(TestCase):
```
This test isn't testing the decorators (so I'd maybe name it something different) and is only testing that your `if` statement is working correctly. It's good to have that test, but the test it replaces also made sure that the correct response code and template are being returned. It might be worth looking at the previous test and seeing how yours can also test those important parts of the code.

pr33thi committed:
Change test_decorators to test_two_factor_check

pr33thi committed:
Change test name

pr33thi committed:
Merge branch 'master' into pv/enable-logout-270986

pr33thi committed:
Add in tests for the two_factor_check function

pr33thi committed:
Clean up tests

pr33thi committed:
Remove unused parts from create_request

Remove prints

pr33thi committed:
Remove unused imports

pr33thi committed:
Remove unneeded parts from create_request

pr33thi committed:
Merge branch 'pv/enable-logout-270986' of https://github.com/dimagi/commcare-hq into pv/enable-logout-270986

pr33thi committed:
Typo fixed

pr33thi committed:
Stickler

pr33thi committed:
Merge branch 'master' into pv/enable-logout-270986

pr33thi committed:
Add tearDown to two factor tests

pr33thi committed:
Uncomment tearDown

pr33thi committed:
Use RequestFactory to create request

proteusvacuum commented on corehq/apps/domain/tests/test_two_factor_check.py:
```diff
@@ -0,0 +1,73 @@
+
+    def test_two_factor_check_without_feature_flag(self):
+        def func_to_check_2fa_on(*_):
+            return 'Function was called!'
```
take a look at: https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called

proteusvacuum reviewed (approved):
Looks good! I'd rewrite that one test using a mock, but non-blocking.

pr33thi committed:
Replacing test function with mock

pr33thi committed:
Replacing test function with mock

Testing mocks and responses. Still have future import commented out

Change string name to bytes for python 3 compatibility

Remove commented line

pr33thi committed:
Testing mocks and responses. Still have future import commented out

pr33thi committed:
Change string name to bytes for python 3 compatibility

pr33thi committed:
Remove commented line

pr33thi committed:
Merge branch 'pv/enable-logout-270986' of https://github.com/dimagi/commcare-hq into pv/enable-logout-270986

pr33thi commented:
@proteusvacuum 
Here are my (squashed) commits since you last reviewed this:
https://github.com/dimagi/commcare-hq/pull/19570/commits/25790854237a5cecf5a7603ad0d7c0e2fb8709e1

pr33thi committed:
Merge branch 'master' into pv/enable-logout-270986

pr33thi committed:
Add TWO_FACTOR_SUPERUSER_ROLLOUT to toggles import

proteusvacuum commented:
:tada:  Thanks for all the back and forth. I think this ended up being much better for it :)
